File: example/1d-linear-convection/weno3/python/06e/flux/rusanov.py
# flux/rusanov.py
"""
Rusanov（Lax-Friedrichs）通量计算器
"""

from cfd_registry import register_component
from .base import InviscidFluxCalculator

@register_component('flux', 'rusanov')
class RusanovFluxCalculator(InviscidFluxCalculator):
    """Rusanov（Lax-Friedrichs）通量"""
    def compute(self, q_face_left, q_face_right, flux):
        for i in range(self.mesh.nnodes):
            u_L = q_face_left[i]
            u_R = q_face_right[i]
            c_L = self.wave_speed
            c_R = self.wave_speed
            F_L = c_L * u_L  # Flux from left state
            F_R = c_R * u_R  # Flux from right state
            Smax = max(abs(c_L), abs(c_R))  # Maximum wave speed
            flux[i] = 0.5 * (F_L + F_R) - 0.5 * Smax * (u_R - u_L)

File: example/1d-linear-convection/weno3/julia/03f/src/flux/rusanov.jl
# src/flux/rusanov.jl

mutable struct RusanovFluxCalculator <: InviscidFluxCalculator
    cfd::Any
    config::Any
    mesh::Mesh
    wave_speed::Float64
    
    function RusanovFluxCalculator(cfd::Any)
        config = cfd.config
        mesh = cfd.domain.mesh
        wave_speed = config.wave_speed
        new(cfd, config, mesh, wave_speed)
    end
end

function compute!(calc::RusanovFluxCalculator, q_face_left::Vector{Float64}, q_face_right::Vector{Float64}, flux::Vector{Float64})
	eq = calc.cfd.equation
    for i in 1:calc.mesh.nnodes
        u_L = q_face_left[i]
        u_R = q_face_right[i]
        c_L = max_wave_speed(eq, u_L)
        c_R = max_wave_speed(eq, u_R)
        F_L = eq_flux(eq, u_L)
        F_R = eq_flux(eq, u_R)
        Smax = max(abs(c_L), abs(c_R))
        flux[i] = 0.5 * (F_L + F_R) - 0.5 * Smax * (u_R - u_L)
    end
end